runtime/compiler/control/CompilationThread.cpp

```diff
@@ -249,7 +259,77 @@ jitSignalHandler(struct J9PortLibrary *portLibrary, U_32 gpType, void *gpInfo, v
 
    return J9PORT_SIG_EXCEPTION_CONTINUE_SEARCH;
    }
+#ifdef COMPRESS_AOT_DATA
+#ifdef J9ZOS390
+#pragma convlit(suspend)
+#endif
+static
+int inflateBuffer(U_8 *buffer, int numberOfBytes, U_8 *outBuffer, int unCompressedSize)
+   {
+   Bytef *in = (Bytef*)buffer;
+   Bytef *out = (Bytef*)outBuffer;
+   z_stream _stream;
+   _stream.zalloc = Z_NULL;
+   _stream.zfree = Z_NULL;
+   _stream.opaque = Z_NULL;
+   _stream.avail_in = 0;
+   _stream.next_in = Z_NULL;
+   int ret = inflateInit(&_stream);
+   if (ret != Z_OK)
+      return DECOMPRESSION_FAILED;
+   _stream.avail_out = unCompressedSize;
+   _stream.next_out = out;
+   _stream.next_in = in;
+   _stream.avail_in = numberOfBytes;
+   ret = inflate(&_stream, Z_NO_FLUSH);
+   switch (ret)
+      {
+      case Z_NEED_DICT:
+         ret = Z_DATA_ERROR;
+      case Z_DATA_ERROR:
+      case Z_MEM_ERROR:
+         (void)inflateEnd(&_stream);
+         return DECOMPRESSION_FAILED;
+      }
+   assert(_stream.avail_out == 0);
+   (void)inflateEnd(&_stream);
+   return 0;
+   }
 
+static
+int deflateBuffer(const U_8 *buffer, int numberOfBytes, U_8 *outBuffer, int level)
+   {
+   z_stream _stream;
+   _stream.zalloc = Z_NULL;
+   _stream.zfree = Z_NULL;
+   _stream.opaque = Z_NULL;
+   int ret=deflateInit(&_stream, level);
+   if (ret != Z_OK)
+      return COMPRESSION_FAILED;
+   // Start deflating
+   _stream.avail_in = numberOfBytes;
+   _stream.next_in = (Bytef*) buffer;
+   _stream.avail_out = numberOfBytes;
+   _stream.next_out = (Bytef*) outBuffer;
+   ret = deflate(&_stream, Z_FINISH);
+   // We passes in the same size buffer for the compressed data.
+   // Return COMPRESSION_FAILED in case we are ending up inflating the data.
+   // Caller of this routine can then decide if they want to retry deflating
+   // Using larger output buffer.
+   if (ret == Z_OK)
```

Comment:
Should the test say "!=" here?

Reply:
Actually this is correct, but misleading. I should have check if ret is `Z_STREAM_END` or not. While deflating the buffer, zlib returns `Z_STREAM_END` it there is nothing left in input buffer stream to compress and return `Z_OK` if there is still some data. If we can not hold compressed data in the output buffer which is of same size as input buffer, there is no point in storing processed data in the case and we should store original uncompressed data. 
 I have commented and changed the condition to correctly reflect this in https://github.com/eclipse/openj9/pull/5926/commits/ed564d03add53c4f02ffb6e81bc3961a7aaddc1e